velox/dwio/common/compression/Compression.cpp

```diff
@@ -194,14 +195,158 @@ uint64_t ZlibDecompressor::decompress(
+        "{} decompression failed, input len is too small: {}",
+        kind_,
+        compressedSize);
+
+    uint32_t decompressedBlockSize =
+        folly::Endian::big(folly::loadUnaligned<uint32_t>(inputPtr));
+    inputPtr += dwio::common::INT_BYTE_SIZE;
+    compressedSize -= dwio::common::INT_BYTE_SIZE;
+    uint32_t remainingOutputSize = uncompressedSize - decompressedTotalSize;
+
+    DWIO_ENSURE_GE(
+        remainingOutputSize,
+        decompressedBlockSize,
+        "{} decompression failed, remainingOutputSize is less than "
+        "decompressedBlockSize, remainingOutputSize: {}, "
+        "decompressedBlockSize: {}",
+        kind_,
+        remainingOutputSize,
+        decompressedBlockSize);
+
+    if (compressedSize <= 0) {
+      break;
+    }
+
+    do {
+      // Check that input length should not be negative.
+      DWIO_ENSURE_GE(
+          compressedSize,
+          dwio::common::INT_BYTE_SIZE,
+          "{} decompression failed, input len is too small: {}",
+          kind_,
+          compressedSize);
+      // Read the length of the next lz4/lzo compressed block.
+      uint32_t compressedBlockSize =
+          folly::Endian::big(folly::loadUnaligned<uint32_t>(inputPtr));
+      inputPtr += dwio::common::INT_BYTE_SIZE;
+      compressedSize -= dwio::common::INT_BYTE_SIZE;
+
+      if (compressedBlockSize == 0) {
+        continue;
+      }
+
+      DWIO_ENSURE_LE(
+          compressedBlockSize,
+          compressedSize,
+          "{} decompression failed, compressedBlockSize is greater than compressedSize, "
+          "compressedBlockSize: {}, compressedSize: {}",
+          kind_,
+          compressedBlockSize,
+          compressedSize);
+
+      // Decompress this block.
+      remainingOutputSize = uncompressedSize - decompressedTotalSize;
+      uint64_t decompressedSize = -1;
+      decompressedSize = decompressInternal(
+          inputPtr,
+          static_cast<int32_t>(compressedBlockSize),
+          outPtr,
+          static_cast<int32_t>(remainingOutputSize));
+
+      DWIO_ENSURE_EQ(
+          decompressedSize,
+          remainingOutputSize,
+          "{} decompression failed, decompressedSize is not equal to remainingOutputSize, "
+          "decompressedSize: {}, remainingOutputSize: {}",
+          kind_,
+          decompressedSize,
+          remainingOutputSize);
+
+      outPtr += decompressedSize;
+      inputPtr += compressedBlockSize;
+      compressedSize -= compressedBlockSize;
+      decompressedBlockSize -= decompressedSize;
+      decompressedTotalSize += decompressedSize;
+    } while (decompressedBlockSize > 0);
+  }
+
+  DWIO_ENSURE_EQ(
+      decompressedTotalSize,
+      uncompressedSize,
+      "{} decompression failed, decompressedTotalSize is not equal to uncompressedSize, "
+      "decompressedTotalSize: {}, uncompressedSize: {}",
+      kind_,
+      decompressedTotalSize,
+      uncompressedSize);
+
+  return decompressedTotalSize - uncompressedSize;
```

Comment:
> For streaming decompressors it should be the current decompressedTotalLength. For other all-at-once decompressors, it should be the difference with the input uncompressedSize and it should be zero, which will be checked by the caller.

By "all-at-once decompressors", did you mean the framed LzoAndLz4Decompressor? Why should it return 0 always? The DWIO_ENSURE_EQ check ensures that decompressedTotalSize - uncompressedSize = 0 and therefore this decompress() always returns 0 for framed LZ4 and LZO decompression. As far as I understand, `Decompressor::decompress()` function should return the number of bytes written to the dest buffer no matter what compression type it is. The PagedInputStream has this callsite:
```
outputBufferLength_ = decompressor_->decompress(
        input,
        remainingLength_,
        outputBuffer_->data(),
        outputBuffer_->capacity());
    *data = outputBuffer_->data();
    *size = static_cast<int32_t>(outputBufferLength_);
    outputBufferPtr_ = outputBuffer_->data() + outputBufferLength_;
```
This is the only callsite, and it suggest it returns the output buffer length, which should not be 0 for most cases. Why would `outputBufferPtr_` be advanced with 0 bytes always?

Reply:
Yes, you are right, I will have a relook at it. I had seen this callsite and verified that tests were all running fine. But right now, it seems parquet compression code is broken (https://github.com/facebookincubator/velox/issues/6890) and I will have a relook at this after a fix for this is introduced.